### PR TITLE
[WIP] Implement live NodeList

### PR DESCRIPTION
**Pull request by KiChjang**
Creating this PR just to see what WPT says for now.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22264)

### CODE REVIEW

File: components/script/dom/bindings/trace.rs [hunk 1]
```diff
@@ -372,6 +373,7 @@ unsafe impl<A: JSTraceable, B: JSTraceable, C: JSTraceable> JSTraceable for (A,
 
 unsafe_no_jsmanaged_fields!(ActiveAttribInfo);
 unsafe_no_jsmanaged_fields!(ActiveUniformInfo);
+unsafe_no_jsmanaged_fields!(Box<dyn Fn(&Node) -> Box<dyn Iterator<Item = DomRoot<Node>>> + 'static>);
```
Comment (emilio) [inline components/script/dom/bindings/trace.rs]: Hmm, how is this safe at all?

File: components/script/dom/document.rs [hunk 2]
```diff
@@ -3853,18 +3844,28 @@ impl DocumentMethods for Document {
...
-                None => return false,
-            };
-            if element.namespace() != &ns!(html) {
-                return false;
-            }
-            element
-                .get_attribute(&ns!(), &local_name!("name"))
-                .map_or(false, |attr| &**attr.value() == &*name)
-        })
+        let doc = self.GetDocumentElement();
```
Comment (emilio) [inline components/script/dom/document.rs]: Why using the document element instead of `self`? A document is a node as well.

**Review by emilio — COMMENTED** [review]
I think we need to come up with something a bit better perf wise... What gecko does is using `nsIMutationObserver` on the root, which is effectively a combination of the `children_changed` / `attribute_changed`/ etc functions.

File: components/script/dom/bindings/trace.rs (hunk 1 above)
Comment (KiChjang) [reply components/script/dom/bindings/trace.rs]: I had my huge doubts as well, and eventually I created a trait to encapsulate this idea instead.

File: components/script/dom/document.rs (hunk 2 above)
Comment (KiChjang) [reply components/script/dom/document.rs]: Yeah, I think there wasn't any reason why the document element needs to be used. Let's update this.

File: components/script/dom/document.rs [hunk 3]
```diff
@@ -3853,18 +3844,28 @@ impl DocumentMethods for Document {
...
+        match doc.r().map(Castable::upcast::<Node>) {
+            Some(doc_node) => NodeList::new_live_list(&self.window, doc_node, move |root| {
+                let name = name.clone();
+                Box::new(
+                    root.traverse_preorder()
```
[the remark this answers is not in the document]
Comment (nox) [reply components/script/dom/document.rs]: The last one needs to be cached.

**Comment by KiChjang** [comment]
Darn, now it looks like I've got some cache invalidation problems...

**Comment by KiChjang** [comment]
So, I think the approach on the latest commit is correct (exactly per @emilio suggested; I'm using a RegisteredObserver on the root to invalidate the cache), but bugs in our current implementation of MutationObservers is causing cache invalidation issues.

File: components/script/dom/node.rs [hunk 4]
```diff
@@ -414,10 +414,13 @@ impl Node {
...
 
     /// Removes the mutation observer for a given node.
-    pub fn remove_mutation_observer(&self, observer: &MutationObserver) {
+    pub fn remove_mutation_observer(&self, obs: &MutationObserver) {
         self.mutation_observers
             .borrow_mut()
-            .retain(|reg_obs| &*reg_obs.observer != observer)
+            .retain(|reg_obs| match reg_obs {
+                RegisteredObserver::Mutation { observer, .. } => &**observer != obs,
+                _ => false,
```
Comment (emilio) [inline components/script/dom/node.rs]: Aren't you removing all of the observers here? Shouldn't this be `true`?

Comment (KiChjang) [reply components/script/dom/node.rs]: Ack, right, lemme push a fix up.

**Comment by pshaughn** [comment]
I apologize for not noticing this PR when I started working on my own solutions to node list liveness issues. I hope you're not upset that I've been duplicating some of your work. If you want to be the one who continues taking this forward, maybe some of what I've written will help you with the subcases you haven't already solved.

**Comment by KiChjang** [comment]
Fixed by #25435.

**Comment by KiChjang** [comment]
@pshaughn No worries! I was asking just to see if anything from this PR was useful to yours. Thanks for picking this up!

**Pull request closed without merging**